addons/stock/models/stock_move.py

```diff
@@ -554,357 +636,251 @@ def check_tracking(self, pack_operation):
-                        if float_compare(lot_qty[lot], 0, precision_rounding=rounding) > 0 and float_compare(move_qty, 0, precision_rounding=rounding) > 0:
-                            qty = min(lot_qty[lot], move_qty)
-                            quants = Quant.quants_get_preferred_domain(qty, move, ops=ops, lot_id=lot, domain=domain, preferred_domain_list=[])
-                            Quant.quants_reserve(quants, move, record)
-                            lot_qty[lot] -= qty
-                            move_qty -= qty
-
-        # Sort moves to reserve first the ones with ancestors, in case the same product is listed in
-        # different stock moves.
-        for move in sorted(moves_to_do, key=lambda x: -1 if ancestors_list.get(x.id) else 0):
-            # then if the move isn't totally assigned, try to find quants without any specific domain
-            if move.state != 'assigned' and not self.env.context.get('reserve_only_ops'):
-                qty_already_assigned = move.reserved_availability
-                qty = move.product_qty - qty_already_assigned
-
-                quants = Quant.quants_get_preferred_domain(qty, move, domain=main_domain[move.id], preferred_domain_list=[])
-                Quant.quants_reserve(quants, move)
-
-        # force assignation of consumable products and incoming from supplier/inventory/production
-        # Do not take force_assign as it would create pack operations
-        if moves_to_assign:
-            moves_to_assign.write({'state': 'assigned'})
-        if not no_prepare:
-            self.check_recompute_pack_op()
+                if not move.move_orig_ids:
+                    if move.procure_method == 'make_to_order':
+                        continue
+                    # Reserve new quants and create move lines accordingly.
+                    available_quantity = self.env['stock.quant'].get_available_quantity(move.product_id, move.location_id)
+                    if available_quantity <= 0:
+                        continue
+                    need = move.product_qty - move.reserved_availability
+                    taken_quantity = move._increase_reserved_quantity(need, available_quantity, move.location_id, strict=False)
+                    if need == taken_quantity:
+                        move.state = 'assigned'
+                    else:
+                        move.state = 'partially_available'
+                else:
+                    # Check what our parents brought and what our siblings took in order to
+                    # determine what we can distribute.
+                    move_lines_in = move.move_orig_ids.filtered(lambda m: m.state == 'done').mapped('pack_operation_ids')
+                    keys_in = ['location_dest_id', 'lot_id', 'result_package_id', 'owner_id']
+                    grouped_move_lines_in = {}
+                    for k, g in groupby(sorted(move_lines_in, key=itemgetter(*keys_in)), key=itemgetter(*keys_in)):
```

Comment:
Bleurk. Isn't it a way to make this a little more readable ? Using itertool or something like this?